File: backend/internal/handlers/finance.go
package handlers

import (
	"fmt"

	"github.com/gofiber/fiber/v2"
	"github.com/inovar/backend/internal/models"
)

// GetFinanceSummary calculates financial stats from requests
func (h *Handler) GetFinanceSummary(c *fiber.Ctx) error {
	var requests []models.Solicitacao
	// Preload OrcamentoItens to have real transaction data
	if err := h.DB.Preload("OrcamentoItens").Find(&requests).Error; err != nil {
		return ServerError(c, err)
	}

	totalRevenue := 0.0
	pendingRevenue := 0.0
	expenses := 0.0

	var expenseRecords []models.Expense
	h.DB.Find(&expenseRecords)
	for _, exp := range expenseRecords {
		expenses += exp.Amount
	}

	transactions := []fiber.Map{}

	for _, req := range requests {
		if req.Status == models.StatusFinalizada {
			totalRevenue += req.ValorOrcamento

			// Transaction for the total OS value
			transactions = append(transactions, fiber.Map{
				"id":          req.ID,
				"date":        req.UpdatedAt, // Use update date as completion date
				"type":        "income",
				"amount":      req.ValorOrcamento,
				"description": fmt.Sprintf("OS #%d - %s", req.Numero, req.ClientName),
				"status":      "paid",
			})
		} else if req.Status != models.StatusCancelada {
			pendingRevenue += req.ValorOrcamento
		}
	}

	// Add individual expense transactions
	for _, exp := range expenseRecords {
		transactions = append(transactions, fiber.Map{
			"id":          exp.ID,
			"date":        exp.Date,
			"type":        "expense",
			"amount":      exp.Amount,
			"description": fmt.Sprintf("[%s] %s", exp.Category, exp.Description),
			"status":      "paid",
		})
	}

	return Success(c, fiber.Map{
		"totalRevenue":   totalRevenue,
		"netProfit":      totalRevenue - expenses,
		"pendingRevenue": pendingRevenue,
		"expenses":       expenses,
		"transactions":   transactions,
	})
}

// ListTransactions returns a list of financial transactions
func (h *Handler) ListTransactions(c *fiber.Ctx) error {
	// Reusing logic from Summary for now, or could be a separate table
	return h.GetFinanceSummary(c)
}

File: infra/pyservice/routes/finance.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List, Optional
import uuid
from datetime import datetime

from database import get_db
from models import Solicitacao, Expense
from pydantic import BaseModel

router = APIRouter()

@router.get("/summary")
def get_summary(db: Session = Depends(get_db)):
    requests = db.query(Solicitacao).all()
    expenses = db.query(Expense).all()

    total_revenue = sum(r.valor_orcamento or 0.0 for r in requests if r.status == "FINALIZADA")
    pending_revenue = sum(r.valor_orcamento or 0.0 for r in requests if r.status not in ["FINALIZADA", "CANCELADA"])
    total_expenses = sum(e.amount for e in expenses)

    transactions = []
    for r in requests:
        if r.status == "FINALIZADA":
            transactions.append({
                "id": r.id,
                "date": r.updated_at,
                "type": "income",
                "amount": r.valor_orcamento,
                "description": f"OS #{r.numero} - {r.client_name}",
                "status": "paid"
            })

    for e in expenses:
        transactions.append({
            "id": e.id,
            "date": e.date,
            "type": "expense",
            "amount": e.amount,
            "description": f"[{e.category}] {e.description}",
            "status": "paid"
        })

    return {
        "data": {
            "totalRevenue": total_revenue,
            "netProfit": total_revenue - total_expenses,
            "pendingRevenue": pending_revenue,
            "expenses": total_expenses,
            "transactions": transactions
        }
    }

@router.get("/expenses")
def list_expenses(db: Session = Depends(get_db)):
    return {"data": db.query(Expense).all()}

@router.post("/expenses")
def create_expense(req: dict, db: Session = Depends(get_db)):
    if "id" not in req:
        req["id"] = str(uuid.uuid4())
    exp = Expense(**req)
    db.add(exp)
    db.commit()
    db.refresh(exp)
    return {"data": exp}
